[Console][Messenger] add `RunCommandMessage` and `RunCommandMessageHandler`

kbond opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | n/a
| License       | MIT
| Doc PR        | todo

Similar to #49813, when using the scheduler it could be useful to execute commands.

## Usage

```php
use Symfony\Component\Console\Exception\RunCommandFailedException;
use Symfony\Component\Console\Messenger\RunCommandMessage;

try {
    $context = $bus->dispatch(new RunCommandMessage('my:command'));

    $context->output; // string - output of command
    $context->exitCode; // int - the exit code
catch(RunCommandFailedException $e) { // if exit code is non-zero or command threw exception
    $e->context->output; // string - output of command
    $e->context->exitCode; // int - the exit code
    $e->getPrevious(); // null|\Throwable exception command threw if applicable
}

// "never" fail
$context = $bus->dispatch(new RunCommandMessage('my:command', throwOnNonSuccess: false, catchExceptions: true));
```

TODO:
- [x] wire up
- [x] tests

kbond committed:
[Console][Messenger] add `RunCommandMessage` and `RunCommandMessageHandler`

kbond commented:
Like https://github.com/symfony/symfony/pull/49813#discussion_r1148413168, maybe `CommandMessage` and `CommandMessageHandler` would be better names?

kbond commented on src/Symfony/Component/Console/Messenger/ExecuteCommandHandler.php:
```diff
@@ -0,0 +1,37 @@
+        $input = new StringInput($message->input);
+        $output = new BufferedOutput();
+
+        $this->application->setCatchExceptions($message->catchExceptions);
```
Should I clone the application here instead of having the service not-shared?

ro0NL replied:
maybe try/finally to modify state on the shared service?

kbond replied:
I don't think I can do this as there's no way to access the current state of `Application::$catchExceptions`.

kbond commented on src/Symfony/Component/Console/Messenger/ExecuteCommandHandler.php:
```diff
@@ -0,0 +1,37 @@
+        $this->application->setCatchExceptions($message->catchExceptions);
+        $this->application->run($input, $output);
+
+        return $output;
```
Any better ideas on what to return here? `BufferedOutput` doesn't contain the exit code.

ro0NL replied:
this would mostly be excuted async isnt it?

ro0NL replied:
we are currently simply logging error output; success output seems irrelevant?

kbond replied:
The idea with returning the output is so it can be captured by a messenger event. My use case is to save the output to the db for later review.

ro0NL replied:
> BufferedOutput doesn't contain the exit code.

maye return a tuple/pair? `return [$extiCode, $output]`

kbond replied:
What about a `ProcessOutput` value object that contains this data to be returned by `ExecuteCommandHandler` and `ExecuteProcessHandler` in #49813?

This could also be added to the exception I describe in https://github.com/symfony/symfony/pull/49814#issuecomment-1486903588.

ro0NL replied:
i assume most cases run async and ignore the return value anyway

we rely on failure/error logging, and IMHO, if such data needs to to be persisted in a SQL database, it should be done by an actual [logger](https://github.com/gupalo/monolog-dbal-logger)

for those cases that consume the return value a simple array sounds fine, but feel free to make it a value object sure :)

kbond commented on src/Symfony/Component/Console/Messenger/ExecuteCommandHandler.php:
```diff
@@ -0,0 +1,37 @@
+        $output = new BufferedOutput();
+
+        $this->application->setCatchExceptions($message->catchExceptions);
+        $this->application->run($input, $output);
```
Should we throw an exception if the exist code isn't `0`?

ro0NL replied:
perhaps only if message.catchExceptions=true we should try-catch Application.run() and log any exception to console channel vs. letting the Application exception bubble otherwise

about the exit code; Process components calls it "mustRun"

kbond replied:
I've made this configurable.

kbond commented:
One thing I think is missing: making the output available if there's an exception. Consider a system that tracks messages and emails an admin on failure. I can attest that this is valuable information.

What about wrapping the caught exception in a new `RunCommandMessageException` that contains the output?

ro0NL commented on src/Symfony/Component/Console/Messenger/ExecuteCommand.php:
```diff
@@ -0,0 +1,29 @@
+    ) {
+    }
+
+    public function __toString(): string
```
see also https://github.com/symfony/symfony/pull/49813#discussion_r1175639793

ro0NL commented on src/Symfony/Component/Console/Messenger/ExecuteCommand.php:
```diff
@@ -0,0 +1,29 @@
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+final class ExecuteCommand implements \Stringable
```
ExecuteConsoleCommand?

kbond replied:
Renamed to RunCommandMessage

kbond commented:
I've updated this PR:
1. Renamed the message to `RunCommandMessage`.
2. Added `RunCommandContext` to be returned from the handler
3. Added `RunCommandFailedException` that's thrown on failure that wraps the original exception and contains `RunCommandContext`
4. Configuration for throwing `RunCommandFailedException` on non-successful exit codes

fabpot commented on src/Symfony/Component/Console/Messenger/RunCommandContext.php:
```diff
@@ -0,0 +1,23 @@
+{
+    public function __construct(RunCommandMessage $message, public readonly int $exitCode, public readonly string $output)
+    {
+        parent::__construct($message->input, $message->throwOnNonSuccess, $message->catchExceptions);
```
Maybe

```suggestion
        parent::__construct($message->input, $message->throwOnFailedCommand, $message->catchExceptions);
```

or

```suggestion
        parent::__construct($message->input, $message->throwOnFailure, $message->catchExceptions);
```

kbond replied:
I went with `throwOnFailure`.

fabpot reviewed (approved):
LGTM, @kbond I've made a rename suggestion.

fabpot commented:
Thank you @kbond.
